scripts/process_uploads.py

```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python3
+"""
+Processes a new batch of samples from the UPLOAD bucket,
+into the MAIN and ARCHIVE buckets using the batch_move_files function. 
+
+Determines the files to be moved through a comparison of the 
+most recently updated CSV and the CSV uploaded in a previous batch run. 
+"""
+
+import os
+import io
+import csv
+from os.path import join, basename
+import subprocess
+from typing import List
+import click
+import hail as hl
+import hailtop.batch as hb
+from google.cloud import storage
+from joint_calling.upload_processor import batch_move_files
+
+
+def samples_from_csv(bucket_name, path):
+    """ Determines list of samples from a given csv file. """
+
+    client = storage.Client()
+    bucket = client.get_bucket(bucket_name)
+    full_path = os.path.join('gs://', path, '*.csv')
+
+    cmd = f'gsutil ls \'{full_path}\''
+    csv_path = subprocess.check_output(cmd, shell=True).decode().strip()
+    prefix = len('gs:///') + len(bucket_name)
+    blob = bucket.get_blob(csv_path[prefix:])
+    data = blob.download_as_string().decode('utf-8')
+    csv_reader = csv.DictReader(io.StringIO(data))
+
+    samples: List[str] = []
+    for row in csv_reader:
+        samples.append(row['sample.sample_name'])
+
+    return samples
+
+
+def determine_samples(
+    upload_bucket: str, upload_path: str, main_bucket: str, previous_batch_path: str
+):
+    """Determine files that should be moved to main vs archive bucket.
+    Determines the difference between the latest CSV file within upload
+    and the CSV file from the most recent batch as the list of samples to
+    be moved.
+
+    Assumptions
+    ==========
+    Only one .csv file exists in the upload bucket and previous batch directory.
+    """
+    all_samples: List[str] = []
+    previous_samples: List[str] = []
+    all_samples = samples_from_csv(upload_bucket, upload_path)
+    previous_samples = samples_from_csv(main_bucket, previous_batch_path)
+
+    samples = set(all_samples) - set(previous_samples)
+
+    cmd = f'gsutil ls \'gs://{upload_path}/*.csv\''
+    curr_csv_file_path = subprocess.check_output(cmd, shell=True).decode().strip()
+
+    return samples, curr_csv_file_path
+
+
+def generate_file_list(samples: List[str]):
+    """ Generate list of expected files, given a list of sampleIDs """
+    main_files: List[str] = []
+    archive_files: List[str] = []
+    for s in samples:
+        main_files.extend([f'{s}.g.vcf.gz', f'{s}.g.vcf.gz.tbi', f'{s}.g.vcf.gz.md5'])
+        archive_files.extend([f'{s}.cram', f'{s}.cram.crai', f'{s}.cram.md5'])
+
+    return main_files, archive_files
+
+
+@click.command()
+@click.option('--batch_number')
+@click.option('--prev_batch')
```

Comment:
Doesn't one imply the other?

Reply:
I think I was anticipating a case where batches may not be processed in order, for whatever reason. In hindsight, that doesn't really work with our workflow. Have amended :)